tools/pythonpkg/duckdb_python.cpp

```diff
@@ -0,0 +1,700 @@
+			vector<Value> args;
+			for (auto &ele : single_query_params) {
+
+				if (ele.is_none()) {
+					args.push_back(Value());
+				} else if (py::isinstance < py::bool_ > (ele)) {
+					args.push_back(Value::BOOLEAN(ele.cast<bool>()));
+				} else if (py::isinstance < py::int_ > (ele)) {
+					args.push_back(Value::BIGINT(ele.cast<int64_t>()));
+				} else if (py::isinstance < py::float_ > (ele)) {
+					args.push_back(Value::DOUBLE(ele.cast<double>()));
+				} else if (py::isinstance < py::str > (ele)) {
+					args.push_back(Value(ele.cast<string>()));
+				} else if (ele.get_type().is(datetime_date)) {
+					throw runtime_error("date parameters not supported yet :/");
+					//args.push_back(Value::DATE(1984, 4, 24));
+				} else if (ele.get_type().is(datetime_datetime)) {
+					throw runtime_error(
+							"datetime parameters not supported yet :/");
+					//args.push_back(Value::TIMESTAMP(1984, 4, 24, 14, 42, 0, 0));
+				} else {
+					throw runtime_error(
+							"unknown param type "
+									+ py::str(ele.get_type()).cast<string>());
+				}
+			}
+			if (prep->n_param != py::len(single_query_params)) {
+				throw runtime_error(
+						"Prepared statments needs " + to_string(prep->n_param)
+								+ " parameters, "
+								+ to_string(py::len(single_query_params))
+								+ " given");
+			}
+			auto res = make_unique<DuckDBPyResult>();
+			res->result = prep->Execute(args);
+			if (!res->result->success) {
+				throw runtime_error(res->result->error);
+			}
+			if (!many) {
+				result = move(res);
+			}
+		}
+		return this;
+	}
+
+	DuckDBPyConnection* append(string name, py::object value) {
+		register_df("__append_df", value);
+		return execute("INSERT INTO \"" + name + "\" SELECT * FROM __append_df");
+	}
+
+	DuckDBPyConnection* register_df(string name, py::object value) {
+		std::ostringstream address;
+		address << (void const*) value.ptr();
+		string pointer_str = address.str();
+
+		// hack alert: put the pointer address into the function call as a string
+		execute(
+				"CREATE OR REPLACE TEMPORARY VIEW \"" + name
+						+ "\" AS SELECT * FROM pandas_scan('" + pointer_str
+						+ "')");
+
+		// keep a reference
+		registered_dfs[name] = value;
+
+		return this;
+	}
+
+	DuckDBPyConnection* unregister_df(string name) {
+		registered_dfs[name] = py::none();
+		return this;
+	}
+
+	DuckDBPyConnection* begin() {
+		execute("BEGIN TRANSACTION");
+		return this;
+	}
+
+	DuckDBPyConnection* commit() {
+		if (connection->context->transaction.IsAutoCommit()) {
+			return this;
+		}
+		execute("COMMIT");
+		return this;
+	}
+
+	DuckDBPyConnection* rollback() {
+		execute("ROLLBACK");
+		return this;
+	}
+
+	py::object getattr(py::str key) {
+		if (key.cast<string>() == "description") {
+			if (!result) {
+				throw runtime_error("no open result set");
+			}
+			return result->description();
+		}
+		return py::none();
+	}
+
+	void close() {
+		connection = nullptr;
+		database = nullptr;
+	}
+
+	// cursor() is stupid
```

Comment:
It's been a couple of years and a global pandemic, but would you care to elaborate @hannes? Super curious about the provenance of this comment 😄 (And the fact that `DuckDBPyConnection` is both a PEP 249 `Connection` _and_ a `Cursor`)